

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>bigwig_analyser &mdash; GelCoverage 1.0.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="GelCoverage 1.0.0 documentation" href="index.html"/>
        <link rel="prev" title="GelCoverageProject Documentation" href="index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> GelCoverage
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="">bigwig_analyser</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#how-to-use-it-from-commandline">How to use it from commandline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#how-to-use-it-from-python">How to use it from python</a></li>
<li class="toctree-l2"><a class="reference internal" href="#different-configurations">Different configurations</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">GelCoverage</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>bigwig_analyser</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/bigwig_analyser.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="bigwig-analyser">
<h1>bigwig_analyser<a class="headerlink" href="#bigwig-analyser" title="Permalink to this headline">¶</a></h1>
<dl class="docutils">
<dt>This script calculates coverage statistics, using a bigwig as input. It has different execution modes.</dt>
<dd><ul class="first last simple">
<li><cite>&#8211;panel</cite>: This mode will calculate the coverage metrics for one panel.</li>
<li><cite>&#8211;gene-list</cite>: This mode will calculate the coverage metrics for a list of genes.</li>
<li><cite>none of above</cite>: This version will calculate the coverage metrics for all genes.</li>
</ul>
</dd>
</dl>
<p>It will output statistics at exon, transcript, gene (by creating a union transcript), chromosome, analysis coding region
(this is panel, gene list or whole coding region) and whole genome. The output format is JSON.</p>
<div class="section" id="how-to-use-it-from-commandline">
<h2>How to use it from commandline<a class="headerlink" href="#how-to-use-it-from-commandline" title="Permalink to this headline">¶</a></h2>
<p>This script is executed in the following way:</p>
<div class="highlight-bash"><div class="highlight"><pre>bigwig_analyser --bw &lt;bigwig.bw&gt; --output &lt;output.json&gt; --config &lt;configuration.config&gt; --wg-regions &lt;non_n_region.bed&gt; --disable-exon-stats
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The configuration file contain the basic configuration to annotate, get panels information, filter transcripts and calculate stats.
This file will be found in <cite>/genomes/resources/genomeref/...</cite>.</p>
</div>
</div>
<div class="section" id="how-to-use-it-from-python">
<h2>How to use it from python<a class="headerlink" href="#how-to-use-it-from-python" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="c1">#Create a dictionary with the configuration</span>
<span class="n">config</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;bw&quot;</span> <span class="p">:</span> <span class="s1">&#39;/path/to/bigwig.bw&#39;</span><span class="p">,</span>
    <span class="s2">&quot;panel&quot;</span> <span class="p">:</span> <span class="bp">None</span><span class="p">,</span>
    <span class="s2">&quot;panel_version&quot;</span><span class="p">:</span> <span class="bp">None</span><span class="p">,</span>
    <span class="s2">&quot;gene_list&quot;</span><span class="p">:</span> <span class="bp">None</span><span class="p">,</span>
    <span class="s2">&quot;coverage_threshold&quot;</span><span class="p">:</span> <span class="mi">15</span><span class="p">,</span>
    <span class="s1">&#39;configuration_file&#39;</span><span class="p">:</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span>
    <span class="s2">&quot;cellbase_species&quot;</span><span class="p">:</span> <span class="s1">&#39;hsapiens&#39;</span><span class="p">,</span>
    <span class="s2">&quot;cellbase_version&quot;</span><span class="p">:</span> <span class="s1">&#39;latest&#39;</span><span class="p">,</span>
    <span class="s2">&quot;cellbase_assembly&quot;</span><span class="p">:</span> <span class="s1">&#39;GRCh37/GRCh38&#39;</span><span class="p">,</span>
    <span class="s2">&quot;cellbase_host&quot;</span><span class="p">:</span> <span class="s1">&#39;10.5.8.201:8080/cellbase-4.5.0-rc&#39;</span><span class="p">,</span>
    <span class="s2">&quot;panelapp_host&quot;</span><span class="p">:</span> <span class="s1">&#39;bioinfo.extge.co.uk/crowdsourcing/WebServices&#39;</span><span class="p">,</span>
    <span class="s2">&quot;panelapp_gene_confidence&quot;</span><span class="p">:</span> <span class="s1">&#39;HighEvidence&#39;</span><span class="p">,</span>
    <span class="s2">&quot;transcript_filtering_flags&quot;</span><span class="p">:</span> <span class="s1">&#39;basic&#39;</span><span class="p">,</span>
    <span class="s2">&quot;transcript_filtering_biotypes&quot;</span><span class="p">:</span> <span class="s1">&#39;IG_C_gene,IG_D_gene,IG_J_gene,IG_V_gene,IG_V_gene,protein_coding,nonsense_mediated_decay,non_stop_decay,TR_C_gene,TR_D_gene,TR_J_gene,TR_V_gene&#39;</span><span class="p">,</span>
    <span class="s2">&quot;exon_padding&quot;</span><span class="p">:</span> <span class="mi">15</span><span class="p">,</span>
    <span class="s2">&quot;wg_stats_enabled&quot;</span><span class="p">:</span> <span class="bp">True</span><span class="p">,</span>
    <span class="s2">&quot;wg_regions&quot;</span><span class="p">:</span> <span class="s1">&#39;/path/to/non_n_regions.bed&#39;</span><span class="p">,</span>
    <span class="s2">&quot;exon_stats_enabled&quot;</span><span class="p">:</span> <span class="bp">False</span><span class="p">,</span>
    <span class="s2">&quot;coding_region_stats_enabled&quot;</span><span class="p">:</span> <span class="bp">True</span>
<span class="p">}</span>

<span class="n">gel_coverage_engine</span> <span class="o">=</span> <span class="n">GelCoverageRunner</span><span class="p">(</span><span class="n">config</span><span class="p">)</span>
<span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">bed</span><span class="p">)</span> <span class="o">=</span> <span class="n">gel_coverage_engine</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
<span class="c1"># Prints output to stdout</span>
<span class="k">with</span> <span class="n">codecs</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="s1">&#39;utf8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output_file</span><span class="p">:</span>
    <span class="n">output_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
        <span class="n">ujson</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span>
            <span class="n">results</span><span class="p">,</span>
            <span class="n">ensure_ascii</span><span class="o">=</span><span class="bp">False</span>
        <span class="p">)</span>
    <span class="p">)</span>
<span class="c1"># Saves the analysed region as a BED file</span>
<span class="n">bed</span><span class="o">.</span><span class="n">saveas</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output</span> <span class="o">+</span> <span class="s2">&quot;.bed&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Please note that this process is highly dependent on the reference genome, use a different assembly o version assembly
will produce wrong results.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">When running an analysis over all genes the resulting JSON will be around 1.5GB, unless you add the flag &#8211;disable-exon-stats,
but in this case you will be missing the exon level statistics and the coverage gaps.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">When running an analysis in panel or gene list mode it might be useful to disable the whole genome statistics to improve performance,
by using the flag &#8211;disable-wg-stats.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Beware that the reference genome and chromosome notation (i.e.: chr prefix or not) should be the same in the input bigwig file and the bed file in wg-regions.</p>
</div>
</div>
<div class="section" id="different-configurations">
<h2>Different configurations<a class="headerlink" href="#different-configurations" title="Permalink to this headline">¶</a></h2>
<p>The program iterates through the bigwig file twice: the first for the analysis of the coding region (panel, gene list or
full) and the second for the analysis of the whole genome.</p>
<p>To run statistics only for a panel from exon level up to panel level, provide a panel (<cite>panel</cite>) and panel
version (<cite>panel_version</cite>) and disable the whole genome statistics (<cite>&#8220;wg_stats_enabled&#8221;: False</cite>), while making sure that
the coding region and the exon level statistics are enabled (<cite>&#8220;coding_region_stats_enabled&#8221;: True</cite> and <cite>&#8220;exon_stats_enabled&#8221;: True</cite>).
Execution time is of some seconds or under a second for small panels. The panel of intellectual disability v1.23 having 1232 genes took 46s.</p>
<p>To run statistics only for a gene list from exon level up to gene list level, provide a gene list (<cite>gene_list</cite>) instead
of panel and panel version and use the same configuration as above.
Execution time similar as the previous</p>
<p>To run statistics only for all genes in the coding region do not provide panel (<cite>panel</cite>) or gene list (<cite>gene_list</cite>),
disable the whole genome statistics (<cite>&#8220;wg_stats_enabled&#8221;: False</cite>) and the exon level statistics (<cite>&#8220;exon_stats_enabled&#8221;: False</cite>)
(the output JSON will be over 1 GB if exon stats are enabled for all genes),
while making sure that the coding region is enabled (<cite>&#8220;coding_region_stats_enabled&#8221;: True</cite>).
Execution time is over 3 hours</p>
<p>To run only whole genome statistics enable <cite>&#8220;wg_stats_enabled&#8221;: True</cite> and disable the coding region statistics
(<cite>&#8220;coding_region_stats_enabled&#8221;: False</cite>). The whole genome analysis might be used in combination with a bed file defining
the region to analyse (e.g.: non N regions) that is to be passed in parameter <cite>&#8220;wg_regions&#8221;: &#8216;/path/to/non_n_regions.bed&#8217;</cite>.
This <cite>wg_regions</cite> can be used to calculate coverage over very specific regions, for instance Cosmic variants if they are set in
a BED file.
Execution time is around 1 hour</p>
<p>Any combination, of the previous should generate a single JSON with all the information.</p>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="index.html" class="btn btn-neutral" title="GelCoverageProject Documentation" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Pablo Riesgo, Pedro Furio, Matthew Parker, Antonio Rueda, Alona Sosinsky.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>